Tests/NIOPosixTests/AsyncChannelBootstrapTests.swift

```diff
@@ -33,43 +33,101 @@ private final class LineDelimiterDecoder: ByteToMessageDecoder {
         }
         return .needMoreData
     }
+
+    func encode(data: ByteBuffer, out: inout ByteBuffer) throws {
+        var data = data
+        out.writeBuffer(&data)
+        out.writeString("\n")
+    }
 }
 
-private final class TLSUserEventHandler: ChannelInboundHandler {
+private final class TLSUserEventHandler: ChannelInboundHandler, RemovableChannelHandler {
     typealias InboundIn = ByteBuffer
     typealias InboundOut = ByteBuffer
+    enum ALPN: String {
+        case string
+        case byte
+        case unknown
+    }
+
+    private var proposedALPN: ALPN?
+
+    init(
+        proposedALPN: ALPN? = nil
+    ) {
+        self.proposedALPN = proposedALPN
+    }
+
+    func handlerAdded(context: ChannelHandlerContext) {
+        guard context.channel.isActive else {
+            return
+        }
+
+        if let proposedALPN = self.proposedALPN {
+            self.proposedALPN = nil
```

Comment:
Since the channel can already be active in handlerAdded we might have to write the "negotiate request" here but we should only write it once. This code here ensures that between handlerAdded and channelActive that we only write it out once and only once we are active.